cmd/server/http/release.go

```diff
@@ -29,70 +30,50 @@ func release(payload *payload, flowSvc *flow.Service) http.HandlerFunc {
 			"service", req.Service,
 			"req", req)
-		var releaseID string
-		switch {
-		case !emptyString(req.Branch):
-			logger.Infof("http: release: service '%s' environment '%s' branch '%s': releasing branch", req.Service, req.Environment, req.Branch)
-			releaseID, err = flowSvc.ReleaseBranch(ctx, flow.Actor{
-				Name:  req.CommitterName,
-				Email: req.CommitterEmail,
-			}, req.Environment, req.Service, req.Branch)
-		case !emptyString(req.ArtifactID):
-			logger.Infof("http: release: service '%s' environment '%s' artifact id '%s': releasing artifact", req.Service, req.Environment, req.ArtifactID)
-			releaseID, err = flowSvc.ReleaseArtifactID(ctx, flow.Actor{
-				Name:  req.CommitterName,
-				Email: req.CommitterEmail,
-			}, req.Environment, req.Service, req.ArtifactID)
-		default:
-			logger.Infof("http: release: service '%s' environment '%s' artifact id '%s' branch '%s': neither branch nor artifact id specified", req.Service, req.Environment, req.ArtifactID, req.Branch)
-			httpinternal.Error(w, "either branch or artifact id must be specified", http.StatusBadRequest)
-			return
-		}
+
+		logger.Infof("http: release: service '%s' environment '%s' artifact id '%s': releasing artifact", req.Service, req.Environment, req.ArtifactID)
+		releaseID, err := flowSvc.ReleaseArtifactID(ctx, flow.Actor{
+			Name:  req.CommitterName,
+			Email: req.CommitterEmail,
+		}, req.Environment, req.Service, req.ArtifactID, intent.NewReleaseArtifact(req.ArtifactID))
+
 		var statusString string
 		if err != nil {
 			if ctx.Err() == context.Canceled {
-				logger.Infof("http: release: service '%s' environment '%s' branch '%s' artifact id '%s': release cancelled", req.Service, req.Environment, req.Branch, req.ArtifactID)
+				logger.Infof("http: release: service '%s' environment '%s' artifact id '%s': release cancelled", req.Service, req.Environment, req.ArtifactID)
 				cancelled(w)
 				return
 			}
 			switch errorCause(err) {
 			case flow.ErrReleaseProhibited:
-				logger.Infof("http: release: service '%s' environment '%s' branch '%s' artifact id '%s': release rejected: branch prohibited in environment: %v", req.Service, req.Environment, req.Branch, req.ArtifactID, err)
-				if req.Branch != "" {
-					httpinternal.Error(w, fmt.Sprintf("branch '%s' cannot be released to environment '%s' due to branch restriction policy", req.Branch, req.Environment), http.StatusBadRequest)
-				} else {
-					httpinternal.Error(w, fmt.Sprintf("artifact '%s' cannot be released to environment '%s' due to branch restriction policy", req.ArtifactID, req.Environment), http.StatusBadRequest)
-				}
+				logger.Infof("http: release: service '%s' environment '%s' artifact id '%s': release rejected: branch prohibited in environment: %v", req.Service, req.Environment, req.ArtifactID, err)
+				httpinternal.Error(w, fmt.Sprintf("artifact '%s' cannot be released to environment '%s' due to branch restriction policy", req.ArtifactID, req.Environment), http.StatusBadRequest)
```

Comment:
I think it might still be valuable to report a specific message to the user based on the intent?
The flow is to me an implementation detail that bubbles out into the error message:

```
$ hamctl release --branch feature --env prod
Error: artifact 'feature-sha-sha' cannot be released to environment 'prod' due to branch restriction policy
```

This can lead to questions like "What is feature-sha-sha?", "I didn't write feature-sha-sha in --branch. What is going on?" etc.

Reply:
did my best to try to do this. Let me know what you think